src/browser-realm.ts

```diff
@@ -6,15 +6,130 @@ import {
     getOwnPropertyDescriptors,
     construct,
     ErrorCreate,
+    WeakMapCreate,
+    isUndefined,
+    ObjectCreate,
+    WeakMapGet,
+    assign,
+    ownKeys,
 } from "./shared";
 
-// caching references to object values that can't be replaced
-// window -> Window -> WindowProperties -> EventTarget
-const blueWindow = globalThis.window;
-const blueDocument = globalThis.document;
-const blueWindowProto = ReflectGetPrototypeOf(blueWindow);
-const blueWindowPropertiesProto = ReflectGetPrototypeOf(blueWindowProto);
-const blueEventTargetProto = ReflectGetPrototypeOf(blueWindowPropertiesProto);
+/**
+ * - Unforgeable prototype references
+ * - Descriptor maps for those unforgeable prototype references
+ */
+interface CachedReferencesRecord {
+    window: WindowProxy;
+    document: Document;
+    WindowProto: object;
+    WindowPropertiesProto: object;
+    EventTargetProto: object;
+    DocumentProto: object;
+    windowDescriptors: PropertyDescriptorMap;
+    WindowProtoDescriptors: PropertyDescriptorMap;
+    WindowPropertiesProtoDescriptors: PropertyDescriptorMap;
+    EventTargetProtoDescriptors: PropertyDescriptorMap;
+};
+
+const cachedGlobalMap: WeakMap<typeof globalThis, CachedReferencesRecord> = WeakMapCreate();
+
+/**
+ * Given a Window reference, extract a set of references that are important
+ * for the sandboxing mechanism, this includes:
+ * - Unforgeable prototypes
+ * - Descriptor maps for those unforgeable prototypes
+ */
+function getCachedReferences(global: typeof globalThis): CachedReferencesRecord {
+    let record: CachedReferencesRecord | undefined = WeakMapGet(cachedGlobalMap, global);
+    if (!isUndefined(record)) {
+        return record;
+    }
+    record = ObjectCreate(null) as CachedReferencesRecord;
+    // caching references to object values that can't be replaced
+    // window -> Window -> WindowProperties -> EventTarget
+    record.window = global.window;
+    record.document = global.document;
+    record.WindowProto = ReflectGetPrototypeOf(record.window);
+    record.WindowPropertiesProto = ReflectGetPrototypeOf(record.WindowProto);
```

Comment:
Is there a more formal name to the object that `window.__proto__` is, like how `window.__proto__.__proto__` is the `EventTarget`?

Reply:
window.__proto__.constructor.name is `Window`, that's why I call it `WindowProto`